File: gfg-weekly-coding-contest-76/Powerfull Integer.py
from typing import List


class Solution:
    # Version 1: Store all changing point and use diff array
    # TC: O(nlogn), SC: O(n)
    '''
    def powerfullInteger(self, n : int, intervals : List[List[int]], k : int) -> int:
        # code here
        if k > n:
            return -1
        point = set()
        diff = {}
        for x, y in intervals:
            point.add(x)
            point.add(y)
            point.add(y + 1)
            if x not in diff:
                diff[x] = 0
            diff[x] += 1
            if y + 1 not in diff:
                diff[y + 1] = 0
            diff[y + 1] -= 1
        
        point = sorted(point)
        current = 0
        ans = -1
        for t in point:
            current += diff.get(t, 0)
            #print(t, current)
            if current >= k:
                ans = t
        return ans
    '''
    
    # Version 2: Improved version 1
    # Store two points for each interval
    # TC: O(nlogn), SC: O(n)
    def powerfullInteger(self, n : int, intervals : List[List[int]], k : int) -> int:
        # code here
        if k > n:
            return -1
        diff = {}
        for x, y in intervals:
            if x not in diff:
                diff[x] = 0
            diff[x] += 1
            if y + 1 not in diff:
                diff[y + 1] = 0
            diff[y + 1] -= 1
        
        current = 0
        ans = -1
        for t in sorted(diff.keys()):
            if diff[t] > 0:
                current += diff[t]
                if current >= k:
                    ans = t
            else:
                if current >= k:
                    ans = t - 1
                current += diff[t]
        return ans
        

File: gfg-weekly-coding-contest-76/Powerfull Integer.java


//User function Template for Java
class Solution{
    // Version 1: Store time point and diff array
    // Store two points for each interval
    // TC: O(nlogn), SC: O(n)
    public static int powerfullInteger(int n,int interval[][],int k)
    {
        TreeMap<Integer, Integer> diff = new TreeMap<>();
        for (int[] pair: interval) {
            diff.put(pair[0], diff.getOrDefault(pair[0], 0) + 1);
            diff.put(pair[1] + 1, diff.getOrDefault(pair[1] + 1 , 0) - 1);
        }
        
        int current = 0;
        int ans = -1;
        for (Map.Entry<Integer, Integer> pair: diff.entrySet()) {
            if (pair.getValue() > 0) {
                current += pair.getValue();
                if (current >= k) {
                    ans = pair.getKey();
                }
            } else {
                if (current >= k) {
                    ans = pair.getKey() - 1;
                } 
                current += pair.getValue();
            }
        }
        return ans;
    }
}